### PR TITLE
Deprecate Stream-related API

### PR DESCRIPTION
**Pull request by jyemin**
Adds:

* TransportSettings abstract class
* NettyTransportSettings subclass of TransportSettings
* A new method on MongoClientSettings to configure TransportSettings

Deprecates:

* MongoClientSettings#streamFactoryFactory
* NettyStreamFactoryFactory
* NettyStreamFactory
* AsynchronousSocketChannelStreamFactory
* AsynchronousSocketChannelStreamFactoryFactory
* BufferProvider
* SocketStreamFactory
* Stream
* StreamFactory
* StreamFactoryFactory
* TlsChannelStreamFactoryFactory

JAVA-5051

### CODE REVIEW

**Review by jyemin — COMMENTED** [review]
Couple of things:

Though JEP 211 indicates that it should be otherwise, and my testing directly with javac confirms the expected behavior, when building the driver I'm getting warnings about imports of classes/interfaces that have been deprecated.  I'm not sure why, or what to do about it.  Options:

1. Fully qualify all references to deprecated types so that the imports can be removed
2. Disable linting for deprecation
3. Live with the warnings until it's time to remove the deprecated types for 5.0

Scala test `The_scala_package_should_mirror_parts_of_com.mongodb.connection_in_org.mongdb.scala.connection` is passing by adding exceptions for the two new classes.  Not sure whether to add Scala wrappers for them instead.

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/annotations/Evolving.java [hunk 1]
```diff
@@ -51,11 +50,6 @@
  *     <td>Not applicable.</td>
  *   </tr>
  *   <tr>
- *     <td>Doing so allows customizing API behavior.</td>
```
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/annotations/Evolving.java:54]: Agreed

File: driver-core/src/main/com/mongodb/connection/NettyTransportSettings.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    private final EventLoopGroup eventLoopGroup;
+    private final Class<? extends SocketChannel> socketChannelClass;
+    private final ByteBufAllocator allocator;
+    private final SslContext sslContext;
+
+    /**
+     * Gets a builder for an instance of {@code NettyStreamFactoryFactory}.
+     * @return the builder
+     */
+    static Builder builder() {
```
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/connection/NettyTransportSettings.java:51]: Copy pasta... agreed

File: driver-core/src/main/com/mongodb/connection/netty/NettyStreamFactoryFactory.java
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/connection/netty/NettyStreamFactoryFactory.java:1]: Agreed

File: driver-core/src/main/com/mongodb/connection/NettyTransportSettings.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                    !(sslContext instanceof ReferenceCountedOpenSslClientContext));
+
+            return this;
+        }
+
+        /**
+         * Build an instance of {@code NettyStreamFactoryFactory}.
+         * @return factory of the netty stream factory
+         */
+        public NettyTransportSettings build() {
```
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/connection/NettyTransportSettings.java]: Done

File: driver-core/src/main/com/mongodb/annotations/Evolving.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/annotations/Evolving.java:54]: Done

File: driver-core/src/main/com/mongodb/connection/NettyTransportSettings.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/connection/NettyTransportSettings.java]: Done

File: driver-core/src/main/com/mongodb/connection/AsynchronousSocketChannelStreamFactoryFactory.java
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/connection/AsynchronousSocketChannelStreamFactoryFactory.java:1]: Done

**Comment by stIncMale** [comment]
When I run `./gradlew clean check jar scalaCheck kotlinCheck testClasses docs -x test -x integrationTest -x spotlessApply` with `javac` 17.0.3, I am still getting lots of warnings. Here is an example:

> Task :driver-core:compileJava
/Users/valentin.kovalenko/Documents/projects/mongo-java-driver/driver-core/src/main/com/mongodb/MongoClientSettings.java:29: warning: [deprecation] StreamFactoryFactory in com.mongodb.connection has been deprecated
import com.mongodb.connection.StreamFactoryFactory;

**Comment by jyemin** [comment]
>  I am still getting lots of warnings.

Yes, I mentioned that [here](https://github.com/mongodb/mongo-java-driver/pull/1195#pullrequestreview-1620955635), and gave several options for how to proceed.

File: driver-core/src/main/com/mongodb/connection/netty/NettyStreamFactoryFactory.java
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/connection/netty/NettyStreamFactoryFactory.java:1]: Done

File: driver-core/src/main/com/mongodb/connection/TransportSettings.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import com.mongodb.annotations.Sealed;
+
+/**
+ * Transport settings for the driver.
+ *
+ * @since 4.11
+ */
+@Sealed
+@Immutable
+public abstract class TransportSettings {
```
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/connection/TransportSettings.java:29]: Going once...

File: driver-scala/src/test/scala/org/mongodb/scala/ApiAliasAndCompanionSpec.scala [hunk 5]
```diff
@@ -207,7 +208,8 @@ class ApiAliasAndCompanionSpec extends BaseSpec {
       "SocketStreamFactory",
       "Stream",
       "SplittablePayload",
-      "TopologyVersion"
+      "TopologyVersion",
+      "TransportSettings"
```
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-scala/src/test/scala/org/mongodb/scala/ApiAliasAndCompanionSpec.scala]: Done

**Comment by stIncMale** [comment]
> Yes, I mentioned that https://github.com/mongodb/mongo-java-driver/pull/1195#pullrequestreview-1620955635, and gave several options for how to proceed.

Sorry, I thought the added `@SuppressWarnings("deprecation")` were supposed to solve that. I like _"2. Disable linting for deprecation_ combined with modified 3: live without the lint until it's time to remove the deprecated types for 5.0 (we'll move the program elements into the `internal` package, remove `@Deprecated` from them, enable the lint back again).

File: driver-core/src/main/com/mongodb/connection/TransportSettings.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/connection/TransportSettings.java:29]: I am good with the name.

**Comment by jyemin** [comment]
I went with option 2 and disabled linting for deprecation

**Comment by stIncMale** [comment]
A fiew more suppressions are needed, as there are still some warnings (or, rather notes). Here is an example:

> Task :driver-sync:compileJava
Note: /Users/valentin.kovalenko/Documents/projects/mongo-java-driver/driver-sync/src/main/com/mongodb/client/internal/MongoClientImpl.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.

**Comment by stIncMale** [comment]
No, I was wrong. `AbstractUnifiedTest` already specifies `@SuppressWarnings("deprecation")` at the class level, and the note is still there. I guess, it's supposed to be like that.

**Review by stIncMale — APPROVED** [review]

**Comment by jyemin** [comment]
> I guess, it's supposed to be like that.

I'm mystified by the behavior of the compiler in this regard.

**Review by rozza — APPROVED** [review]
So whats next after this PR?

As internally we continue to use `StreamFactoryFactory` is there a ticket, I'm interested in what the alternative will look like.

**Comment by jyemin** [comment]
> So whats next after this PR?

Similar to what we did with the Operation layer in 3.x.  We'll remove the two deprecated methods on MongoClientSettings, and then move all thee deprecated types to `com.mongodb.internal.connection`.  After that we can see if we want to refactor anything.  One obvious thing is to get rid of StreamFactoryFactory and just use StreamFactory directly.  Another is to get rid of `Stream#supportsAdditionalTimeout` as that was added only due to `Stream` being a public type. We'll also need to deal with the removal of `ServerAddress#getSocketAddress`, which we can do once `Stream` is no longer public.  There may be other simplifications, but these are the ones I'm aware of.

**Pull request merged**